Fix a precision issue in `abstract_iteration`

martinholters opened the pull request:
The first loop is left before `valtype` and `statetype` are updated from the current `stateordonet`, so the second loop should first process it before obtaining a new one instead of vice versa.

Fixes #41022.

martinholters committed:
Fix a precision issue in `abstract_iteration`

If the first loop exits in the first iteration, the `statetype` is still
`Bottom`. In that case, the new `stateordonet` needs to be determined
with the two-arg version of `iterate` again.

Fixes #41022.

martinholters commented:
Ah, no, this isn't it. We need to compute a new `stateordonet` based on the const-widened `statetype` first. We just need to be careful in case the first loop never sets a `statetype` to not use `Union{}` as the argument type then.

Keno commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -862,8 +862,12 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
     statetype = widenconst(statetype)
     valtype = widenconst(valtype)
-    while valtype !== Any
+    if statetype === Bottom
```
Is the issue here when skipping the loop above? Maybe we should directly widen stateordonet then? Otherwise we're redoing this call here which duplicates work that was already done above.

martinholters replied:
Widening `stateordonet` may not be sufficient, we need to compute it from the widened `statetype` as input to `iterate`. However, I also dislike the duplicaton of work here. (Which, btw, we also did before as the loop started with computing a new `stateordonet`.) However, I think I have found an alternative that avoids the duplication at the cost of replacing a `<:` with `⊑`. Will push momentarily; let me know what you think.

martinholters committed:
Slightly simplify `abstract_iteration`

martinholters commented:
Bump. @Keno, mind taking another look?

martinholters commented:
Bump. Any objections against merging this?

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -860,17 +860,15 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
-    statetype = widenconst(statetype)
     valtype = widenconst(valtype)
     while valtype !== Any
-        stateordonet = abstract_call_known(interp, iteratef, nothing, Any[Const(iteratef), itertype, statetype], sv).rt
```
This is the only line that should be moved, though I think we can rearrange some others a bit to sometimes skip work that is unnecessary.
```diff
diff --git a/base/compiler/abstractinterpretation.jl b/base/compiler/abstractinterpretation.jl
index 296a30cdb2..0da2ce68b3 100644
--- a/base/compiler/abstractinterpretation.jl
+++ b/base/compiler/abstractinterpretation.jl
@@ -860,25 +860,28 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
     end
     # From here on, we start asking for results on the widened types, rather than
     # the precise (potentially const) state type
-    statetype = widenconst(statetype)
     valtype = widenconst(valtype)
-    while valtype !== Any
-        stateordonet = abstract_call_known(interp, iteratef, nothing, Any[Const(iteratef), itertype, statetype], sv).rt
-        stateordonet = widenconst(stateordonet)
-        nounion = typesubtract(stateordonet, Nothing, 0)
-        if !isa(nounion, DataType) || !(nounion <: Tuple) || isvatuple(nounion) || length(nounion.parameters) != 2
-            valtype = Any
-            break
-        end
-        if nounion.parameters[1] <: valtype && nounion.parameters[2] <: statetype
-            if typeintersect(stateordonet, Nothing) === Union{}
-                # Reached a fixpoint, but Nothing is not possible => iterator is infinite or failing
-                return Any[Bottom], nothing
+    statetype = widenconst(statetype)
+    if valtype !== Any
+        while true
+            stateordonet = widenconst(stateordonet)
+            nounion = typesubtract(stateordonet, Nothing, 0)
+            if !isa(nounion, DataType) || !(nounion <: Tuple) || isvatuple(nounion) || length(nounion.parameters) != 2
+                valtype = Any
+                break
             end
-            break
+            if nounion.parameters[1] <: valtype && nounion.parameters[2] <: statetype
+                if typeintersect(stateordonet, Nothing) === Union{}
+                    # Reached a fixpoint, but Nothing is not possible => iterator is infinite or failing
+                    return Any[Bottom], nothing
+                end
+                break
+            end
+            valtype = tmerge(valtype, nounion.parameters[1])
+            valtype === Any && break
+            statetype = tmerge(statetype, nounion.parameters[2])
+            stateordonet = abstract_call_known(interp, iteratef, nothing, Any[Const(iteratef), itertype, statetype], sv).rt
         end
-        valtype = tmerge(valtype, nounion.parameters[1])
-        statetype = tmerge(statetype, nounion.parameters[2])
     end
     push!(ret, Vararg{valtype})
     return ret, nothing
```

martinholters replied:
No, that leads to the following unsoundness:
```julia
julia> f() = (1:40...,)
f (generic function with 1 method)

julia> @code_typed f()
CodeInfo(
1 ─     Core._apply_iterate(Base.iterate, Core.tuple, $(QuoteNode(1:40)))::Union{}
└──     unreachable
) => Union{}
```
In this case, the first loop is terminates because `MAX_TUPLE_SPLAT` is exceeded. Then `statetype`, `valtype` and `stateordonet` all are `Const`s. Next we widen them to `Int`, `Int`, and `Tuple{Int,Int}`, respectively. So `nounion == Tuple{Int,Int}` also thus `nounion.parameters[1] <: valtype && nounion.parameters[2] <: statetype`. But at the same time, `typeintersect(stateordonet, Nothing) === Union{}`, so a `nothing`-free fixpoint is declared erroneously. That's why in my version, I don't `widenconst` the `statetype` and use `⊑`, so that in the first iteration `nounion.parameters[2] ⊑ statetype == Int ⊑ Const(n) == false`. In successive iterations, `statetype` will always be a proper type, so we could use `<:` then, but the overhead of `⊑` in the both-arguments-are-types case seems negligible here so I chose not to worry about it.

I think somewhere in the `abstractarray` tests we do a splat of a looong range which would have caught this, but I'll add a more direct test while at it.

vtjnash replied:
Ah, that is true. I marked it as a bugfix, since it appeared to me that we might end up skipping including the first `valtype` (from `iterate(foo)`) into the final result, as we begin this loop with discarding that result and computing `iterate(foo, state)`

martinholters replied:
> I marked it as a bugfix, since it appeared to me that we might end up skipping including the first valtype (from iterate(foo)) into the final result

...but we basically repeated that `iterate` (with widened, but not updated state), so this was indeed not a correctness issue ... right?

vtjnash replied:
yeah, mostly we do repeat the same, but it appears to get specifically confused by that particular case, but perhaps it doesn't get this wrong. Since we start with statetype===Bottom, it will conclude that the result is Bottom, and then luckily we will replace that with Any, but it is not great.

```
iterate(::Foo) = rand(Bool) ? nothing : (nothing, 1)
iterate(::Foo, i) = i < 16 ? (i, i + 1) : nothing
```

vtjnash commented:
Thanks for solving this. It looks like it is actually a bugfix so we should backport it, but that only the iterate function needs to be moved.

martinholters commented:
I don't think this is a bugfix in the sense that inference was unsound before. At least I could not come up with a case, although that may not be too strong an indicator... On the other hand, the fact that you proposed an alternative which did end up being unsound is a strong indicator that this is quite tricky to get right, so I'm not sure backporting is a good idea.

martinholters committed:
Add inference soundness test for splatting long range

Explicitly test that inference produces a sound (and reasonably precise)
result when splatting an iterator (in this case a long range) that
allows constant-propagation up to the `MAX_TUPLE_SPLAT` limit.

martinholters commented:
> but that only the iterate function needs to be moved (the other changes just make it slower / less accurate)

If you mean that we should not do the removal of `statetype = widenconst(statetype)` and use of `⊑` instead of `<:`, then no, these changes are needed for soundness (see explanation in response to your inline comment above).

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -860,17 +860,15 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
     # From here on, we start asking for results on the widened types, rather than
     # the precise (potentially const) state type
-    statetype = widenconst(statetype)
     valtype = widenconst(valtype)
```
```suggestion
    valtype = Bottom
```

martinholters replied:
True; I was basically `tmerge`ing this with itself after the first iteration, so could indeed just start with `Bottom` here to maybe save a bit. But more importantly, that will ensure we don't declare a fixpoint in the first iteration, so we could go back to widening the state here and using `<:` in the loops break condition. Let me play with that locally and see whether I missed anything in that reasoning.

vtjnash replied:
`valtype` is the type of the previous element, not the type of the current element. It was incorrect (though conservative) to include it

martinholters replied:
Ah, true.

martinholters committed:
Optimize `abstract_iteration` a little bit

Avoid recomputation of a `widenconst` and initialize `valtype` and
`statetype` to `Bottom` before second loop to simplify a `⊑`to a `<:`
there.

martinholters commented:
Prodding at this, I've found another case where we could consider improving precision:
```julia
julia> struct NoneOrFailOnSecond{T}
           itr::T
       end

julia> Base.iterate(itr::NoneOrFailOnSecond) = iterate(itr.itr)

julia> f_splat(x) = (x...,)
f_splat (generic function with 1 method)

julia> f_splat(NoneOrFailOnSecond([]))
()

julia> f_splat(NoneOrFailOnSecond([1]))
ERROR: MethodError: no method matching iterate(::NoneOrFailOnSecond{Vector{Int64}}, ::Int64)
[...]

julia> @code_typed f_splat(NoneOrFailOnSecond([1]))
CodeInfo(
1 ─ %1 = Core._apply_iterate(Base.iterate, Core.tuple, x)::Tuple
└──      return %1
) => Tuple
```
(Same for master and this PR in its current form.) Conceivably, this could be inferred as `Tuple{}`. It should also be possible to construct a case where it would be possible to infer that either a fixed (non-zero) number of elements is produced or the iteration fails. We would need to check `nounion == Bottom` and then act accordingly. But I doubt it's worth it. Such iterators look rather pathological and defensively producing too wide a result seems better than getting something wrong and returning too narrow a result, even if only for bizarre edge cases.

So I'd say leave that as is (for now, at least). Opinions?

vtjnash commented:
I think we can easily handle that, by choosing not to widen valtype, and avoiding adding Vararg{Bottom} to the array

Another case we can improve is to model the implicit typeassert call, by adding a `tmeet(nounion, Tuple{Any, Any})` call before the checks for whether we have an Tuple of 2 fields

martinholters committed:
Improve `abstract_iteration` for failing and infinte iterators

martinholters commented on test/compiler/inference.jl:
```diff
@@ -2872,9 +2872,24 @@ partial_return_2(x) = Val{partial_return_1(x)[2]}
 f_splat(x) = (x...,)
 @test Base.return_types(f_splat, (Pair{Int,Int},)) == Any[Tuple{Int, Int}]
+@test Base.return_types(f_splat, (UnitRange{Int},)) == Any[Tuple{Vararg{Int}}]
+struct Itr41839_1 end # empty or infinite
+Base.iterate(::Itr41839_1) = rand(Bool) ? (nothing, nothing) : nothing
+Base.iterate(::Itr41839_1, ::Nothing) = (nothing, nothing)
+@test Base.return_types(f_splat, (Itr41839_1,)) == Any[Tuple{}]
```
And we finally have a bugfix! This gives `Union{}` on master.

vtjnash replied:
yay! that is quite the iteration space

martinholters commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -860,27 +862,43 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
-        stateordonet = widenconst(stateordonet)
-        nounion = typesubtract(stateordonet, Nothing, 0)
-        if !isa(nounion, DataType) || !(nounion <: Tuple) || isvatuple(nounion) || length(nounion.parameters) != 2
+        nounion = typeintersect(typesubtract(stateordonet, Nothing, 0), Tuple{Any,Any})
```
```suggestion
        nounion = typeintersect(stateordonet, Tuple{Any,Any})
```
This should work just as well, right?

vtjnash replied:
LGTM

martinholters commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -860,27 +862,43 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
         end
         if nounion.parameters[1] <: valtype && nounion.parameters[2] <: statetype
+            # reached a fixpoint or iterator failed/gave invalid answer
             if typeintersect(stateordonet, Nothing) === Union{}
```
```suggestion
            if !(Nothing <: stateordonet)
```
As `Nothing` is a leaf type, this should be equal but more efficient?

vtjnash replied:
It should be essential identical (since typeintersect can check that property too), so not worth worrying about the difference too much

martinholters commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -860,27 +862,43 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
+        if nounion === Union{}
+            # handle iterator failing or giving an invalid answer below
+            nounion = Tuple{Union{}, Union{}}
+        elseif !isa(nounion, DataType) || !(nounion <: Tuple) || isvatuple(nounion) || length(nounion.parameters) != 2
```
```suggestion
        elseif !isa(nounion, DataType)
```
Should be sufficient with the `typeintersect` above, but in corner cases it might give funnily too wide types which then might make these extra checks necessary?

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -860,27 +862,43 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
+        nounion = typeintersect(typesubtract(stateordonet, Nothing, 0), Tuple{Any,Any})
+        if nounion === Union{}
+            # handle iterator failing or giving an invalid answer below
+            nounion = Tuple{Union{}, Union{}}
```
subtyping may have some difficulty with this type. it is generally best to normalize away from it, not towards it

martinholters replied:
Good point. That was an attempt to avoid repeating the code for handling infinite iterators below. I'll structure this differently to avoid this "denormalization".

martinholters commented:
@vtjnash anything left I should address or good to go?
